## Tratamento de Exceções

# Definição

```
Uma maneira de tentar contornar esses imprevistos é realizar o tratamento dos locais no código que podem vir a lançar possíveis exceções, como por exemplo, campo de consulta a banco de dados, locais em que há divisões, consulta a arquivos de propriedades ou arquivos dentro do próprio computador.

try
{
  //trecho de código que pode vir a lançar uma exceção
}
catch(tipo_exceçao_1 e)
{
  //ação a ser tomada
}
catch(tipo_exceçao_2 e)
{
  //ação a ser tomada
}
catch(tipo_exceçao_n e)
{
  //ação a ser tomada
}
```

# Exceções comuns

```
A linguagem Java possui um mecanismo especial para o tratamento de erros que possam ocorrer em tempo de execução do programa. Diferentemente de outras linguagens, o surgimento de um erro ocasiona a interrupção imediata do programa, porém em Java podemos tratar esta situação de erro de uma forma adequada e evitando, assim, a interrupção do programa.

package material.excecao;

/**
 * Classe utilizada para demonstrar o bloco try / catch.
 */
public class ExemploExcecao {
  public static void main(String[] args) {
    try {
      /* Trecho de código no qual uma
       * exceção pode acontecer.
       */
    } catch (Exception ex) {
      /* Trecho de código no qual uma
       * exceção do tipo "Exception" será tratada.
       */
    }
  }
}
```
# Divisão por zero

```
for (int i = 1; i <= 5; i++) {
        System.out.println(i);
        int a = i / 0;
        System.out.println(a);
    }
```

# Conversão de tipos de dados inválidos

```
Ignorar a exceção: Dessa forma, o programa terminará anormalmente ou produzirá resultados incorretos;

Atribuir a resolução do problema ao usuário: Como o usuário não é programador, é muito provável que ele não saberá solucionar o problema;

Tratar o erro no programa: O programador desenvolve a aplicação de forma que, quando o erro ocorrer, é acionado um bloco de código responsável pelo tratamento do erro.
```

# Acessar uma posição inválida em um arranjo

```
public class Main
{
    public static void main(String[] args) {
        int[] arr = {4, 3, 2, 3, 1};
        int index = 6;
 
        System.out.println("The element at index " + index + " is " + arr[index]);
    }
}
```

# Acessar uma String nula

```
public class MyClass {
    public static void main(String args[]) {
        String str1 = null;
        String str2 = "Some text";
        if (str1 == null)
            System.out.println("str1 is a null string");
        else
            System.out.println("str1 is not a null string");
            
        if (str2 == null)
            System.out.println("str2 is a null string");
        else
            System.out.println("str2 is not a null string");
    }
}
```

## Bloco para capturar uma exceção

```
Tentar abrir um arquivo que não existe.

Tentar fazer consulta a um banco de dados que não está disponível.

Tentar escrever algo em um arquivo sobre o qual não se tem permissão de escrita.

Tentar conectar em servidor inexistente.

public static void main(String args[])
{
  String frase = null;
  String novaFrase = null;
  try
  {
    novaFrase = frase.toUpperCase();
  }
  catch(NullPointerException e) //CAPTURA DA POSSÍVEL exceção.
  {
    //TRATAMENTO DA exceção
    System.out.println("O frase inicial está nula,
    para solucional tal o problema, foi lhe atribuito um valor default.");
    frase = "Frase vazia";
    novaFrase = frase.toUpperCase();
  }
  System.out.println("Frase antiga: "+frase);
  System.out.println("Frase nova: "+novaFrase);
}
```

## Bloco para capturar diferentes exceções

```
public class Teste {
public static void main(String[] args){

    try {  
        int a = 0;
        int b = 0;
        int res = a/b;
        System.out.println(res);
    }  
    catch (Exception e) {   
        System.out.println("Erro.");  
    }  
    finally {  
        System.out.println("Finally");
    }

    }

    }

```

## Lançar uma exceção

```

Uma maneira de tentar contornar esses imprevistos é realizar o tratamento dos locais no código que podem vir a lançar possíveis exceções, como por exemplo, campo de consulta a banco de dados, locais em que há divisões, consulta a arquivos de propriedades ou arquivos dentro do próprio computador.

try
{
  //trecho de código que pode vir a lançar uma exceção
}
catch(tipo_exceçao_1 e)
{
  //ação a ser tomada
}
catch(tipo_exceçao_2 e)
{
  //ação a ser tomada
}
catch(tipo_exceçao_n e)
{
  //ação a ser tomada
}

```













